// @codious/ngx-google-places-autocomplete/bundles/codious-ngx-google-places-autocomplete.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@codious/ngx-google-places-autocomplete', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = global || self, factory((global.codious = global.codious || {}, global.codious['ngx-google-places-autocomplete'] = {}), global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (exports, core, forms, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                __createBinding(exports, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG = new core.InjectionToken('NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG');

    var NgxGooglePlacesAutocompleteControlValueAccessor = /** @class */ (function () {
        function NgxGooglePlacesAutocompleteControlValueAccessor(injector) {
            this.injector = injector;
        }
        Object.defineProperty(NgxGooglePlacesAutocompleteControlValueAccessor.prototype, "control", {
            get: function () {
                return this.formControl || this.controlContainer.control.get(this.formControlName);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(NgxGooglePlacesAutocompleteControlValueAccessor.prototype, "controlContainer", {
            get: function () {
                return this.injector.get(forms.ControlContainer);
            },
            enumerable: false,
            configurable: true
        });
        NgxGooglePlacesAutocompleteControlValueAccessor.prototype.registerOnChange = function (fn) {
            this.formControlDirective.valueAccessor.registerOnChange(fn);
        };
        NgxGooglePlacesAutocompleteControlValueAccessor.prototype.registerOnTouched = function (fn) {
            this.formControlDirective.valueAccessor.registerOnTouched(fn);
        };
        NgxGooglePlacesAutocompleteControlValueAccessor.prototype.setDisabledState = function (isDisabled) {
            this.formControlDirective.valueAccessor.setDisabledState(isDisabled);
        };
        NgxGooglePlacesAutocompleteControlValueAccessor.prototype.writeValue = function (obj) {
            this.formControlDirective.valueAccessor.writeValue(obj);
        };
        return NgxGooglePlacesAutocompleteControlValueAccessor;
    }());
    NgxGooglePlacesAutocompleteControlValueAccessor.propDecorators = {
        formControl: [{ type: core.Input }],
        formControlName: [{ type: core.Input }],
        formControlDirective: [{ type: core.ViewChild, args: [forms.FormControlDirective, { static: true },] }]
    };

    var NgxGooglePlacesAutocompleteComponent = /** @class */ (function (_super) {
        __extends(NgxGooglePlacesAutocompleteComponent, _super);
        function NgxGooglePlacesAutocompleteComponent(changeDetectorRef, elementRef, injector, ngxGooglePlacesAutocompleteConfig) {
            var _this = _super.call(this, injector) || this;
            _this.changeDetectorRef = changeDetectorRef;
            _this.elementRef = elementRef;
            _this.ngxGooglePlacesAutocompleteConfig = ngxGooglePlacesAutocompleteConfig;
            _this.index = -1;
            _this.predictions = [];
            _this.show = false;
            _this.placeholder = 'Enter a location';
            _this.scriptLoaded = new core.EventEmitter();
            _this.initialize().then(function () { return _this.service = new google.maps.places.AutocompleteService(); });
            return _this;
        }
        NgxGooglePlacesAutocompleteComponent.prototype.onBlur = function () {
            this.reset(true);
        };
        NgxGooglePlacesAutocompleteComponent.prototype.onClick = function () {
            this.control.setValue(null);
        };
        NgxGooglePlacesAutocompleteComponent.prototype.onFocus = function () {
            this.reset(true, true);
        };
        NgxGooglePlacesAutocompleteComponent.prototype.onKeyDown = function (event) {
            if (!this.show)
                return;
            switch (event.key) {
                case 'ArrowDown':
                    this.index += 1;
                    if (this.index >= this.predictions.length)
                        this.index = this.predictions.length - 1;
                    break;
                case 'ArrowUp':
                    this.index -= 1;
                    if (this.index < 0)
                        this.index = 0;
                    break;
                case 'Escape':
                    this.reset(true);
                    break;
                case 'Enter':
                    this.index !== -1 ? this.onMouseDown(this.predictions[this.index]) : this.reset(true);
                    break;
            }
        };
        NgxGooglePlacesAutocompleteComponent.prototype.onModelChange = function (input) {
            var _this = this;
            if (!input) {
                this.reset();
                return;
            }
            var request = Object.assign({ input: input }, this.ngxGooglePlacesAutocompleteConfig.options);
            this.service.getPlacePredictions(request, function (predictions, status) {
                if (status !== google.maps.places.PlacesServiceStatus.OK) {
                    _this.reset();
                    return;
                }
                _this.predictions = predictions;
                _this.show = true;
                _this.changeDetectorRef.detectChanges();
            });
        };
        NgxGooglePlacesAutocompleteComponent.prototype.onMouseDown = function (prediction, event) {
            this.control.setValue(prediction.description, { emitViewToModelChange: false });
            if (event)
                this.dispatchEvent(event);
            this.reset(true);
        };
        NgxGooglePlacesAutocompleteComponent.prototype.dispatchEvent = function (event) {
            var customEvent;
            if (window.CustomEvent) {
                customEvent = new CustomEvent('submit', { bubbles: true, cancelable: true, detail: event });
            }
            else {
                customEvent = document.createEvent('CustomEvent');
                customEvent.initCustomEvent('submit', true, true, { data: event });
            }
            this.elementRef.nativeElement.firstElementChild.firstElementChild.form.dispatchEvent(customEvent);
        };
        NgxGooglePlacesAutocompleteComponent.prototype.initialize = function () {
            var _a;
            return __awaiter(this, void 0, void 0, function () {
                var _this = this;
                return __generator(this, function (_b) {
                    if ((_a = window.google) === null || _a === void 0 ? void 0 : _a.maps)
                        return [2 /*return*/, new Promise(function (resolve) { return resolve(); })];
                    return [2 /*return*/, new Promise(function (resolve, reject) {
                            window['ngxGooglePlacesAutocompleteCallback'] = function () { _this.scriptLoaded.emit(true); resolve(); };
                            var script = document.createElement('script');
                            script.async = true;
                            script.defer = true;
                            script.onerror = function (err) { return reject(err); };
                            script.src = "https://maps.googleapis.com/maps/api/js?callback=ngxGooglePlacesAutocompleteCallback&key=" + _this.ngxGooglePlacesAutocompleteConfig.key + "&language=" + _this.ngxGooglePlacesAutocompleteConfig.language + "&libraries=" + _this.ngxGooglePlacesAutocompleteConfig.libraries + "&region=" + _this.ngxGooglePlacesAutocompleteConfig.region;
                            script.type = 'text/javascript';
                            document.body.appendChild(script);
                        })];
                });
            });
        };
        NgxGooglePlacesAutocompleteComponent.prototype.reset = function (keep, show) {
            if (keep === void 0) { keep = false; }
            if (show === void 0) { show = false; }
            this.index = -1;
            if (!keep)
                this.predictions = [];
            this.show = show && this.predictions.length > 0;
        };
        return NgxGooglePlacesAutocompleteComponent;
    }(NgxGooglePlacesAutocompleteControlValueAccessor));
    NgxGooglePlacesAutocompleteComponent.decorators = [
        { type: core.Component, args: [{
                    providers: [{ provide: forms.NG_VALUE_ACCESSOR, useExisting: NgxGooglePlacesAutocompleteComponent, multi: true }],
                    selector: 'ngx-google-places-autocomplete',
                    template: "<div>\n  <input autocomplete=\"off\" (blur)=\"onBlur()\" (focus)=\"onFocus()\" [formControl]=\"control\" (keydown)=\"onKeyDown($event)\" (ngModelChange)=\"onModelChange($event)\" [placeholder]=\"placeholder\" type=\"text\" />\n  <button (click)=\"onClick()\" *ngIf=\"control.value\" type=\"button\">&times;</button>\n</div>\n<ul *ngIf=\"show\">\n  <li [class.selected]=\"i === index\" [innerHTML]=\"prediction.innerHTML\" (mousedown)=\"onMouseDown(prediction, $event)\" *ngFor=\"let prediction of predictions | highlight; let i = index\"></li>\n</ul>"
                },] }
    ];
    NgxGooglePlacesAutocompleteComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef },
        { type: core.ElementRef },
        { type: core.Injector },
        { type: undefined, decorators: [{ type: core.Inject, args: [NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG,] }] }
    ]; };
    NgxGooglePlacesAutocompleteComponent.propDecorators = {
        placeholder: [{ type: core.Input }],
        scriptLoaded: [{ type: core.Output }]
    };

    var NgxGooglePlacesAutocompletePipe = /** @class */ (function () {
        function NgxGooglePlacesAutocompletePipe() {
        }
        NgxGooglePlacesAutocompletePipe.prototype.transform = function (predictions) {
            var e_1, _b;
            var _a;
            try {
                for (var predictions_1 = __values(predictions), predictions_1_1 = predictions_1.next(); !predictions_1_1.done; predictions_1_1 = predictions_1.next()) {
                    var prediction = predictions_1_1.value;
                    prediction.innerHTML = prediction.description;
                    for (var i = 0; i < ((_a = prediction.matched_substrings) === null || _a === void 0 ? void 0 : _a.length); i += 1) {
                        var offset = prediction.matched_substrings[i].offset + i * 17;
                        var length = prediction.matched_substrings[i].length + offset;
                        prediction.innerHTML =
                            prediction.innerHTML.slice(0, offset) + "<strong>" + prediction.innerHTML.slice(offset, length) + "</strong>" + prediction.innerHTML.slice(length);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (predictions_1_1 && !predictions_1_1.done && (_b = predictions_1.return)) _b.call(predictions_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return predictions;
        };
        return NgxGooglePlacesAutocompletePipe;
    }());
    NgxGooglePlacesAutocompletePipe.decorators = [
        { type: core.Pipe, args: [{ name: 'highlight' },] }
    ];

    var NgxGooglePlacesAutocompleteModule = /** @class */ (function () {
        function NgxGooglePlacesAutocompleteModule() {
        }
        NgxGooglePlacesAutocompleteModule.forRoot = function (ngxGooglePlacesAutocompleteConfig) {
            return {
                ngModule: NgxGooglePlacesAutocompleteModule,
                providers: [{ provide: NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG, useValue: ngxGooglePlacesAutocompleteConfig }]
            };
        };
        return NgxGooglePlacesAutocompleteModule;
    }());
    NgxGooglePlacesAutocompleteModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [NgxGooglePlacesAutocompleteComponent, NgxGooglePlacesAutocompletePipe],
                    exports: [NgxGooglePlacesAutocompleteComponent],
                    imports: [common.CommonModule, forms.ReactiveFormsModule]
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG = NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG;
    exports.NgxGooglePlacesAutocompleteComponent = NgxGooglePlacesAutocompleteComponent;
    exports.NgxGooglePlacesAutocompleteModule = NgxGooglePlacesAutocompleteModule;
    exports.ɵa = NgxGooglePlacesAutocompleteControlValueAccessor;
    exports.ɵb = NgxGooglePlacesAutocompletePipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
