// @codious/ngx-google-places-autocomplete/esm2015/lib/ngx-google-places-autocomplete.control-value-accessor.js
import { Input, ViewChild } from '@angular/core';
import { ControlContainer, FormControlDirective } from '@angular/forms';
export class NgxGooglePlacesAutocompleteControlValueAccessor {
    constructor(injector) {
        this.injector = injector;
    }
    get control() {
        return this.formControl || this.controlContainer.control.get(this.formControlName);
    }
    get controlContainer() {
        return this.injector.get(ControlContainer);
    }
    registerOnChange(fn) {
        this.formControlDirective.valueAccessor.registerOnChange(fn);
    }
    registerOnTouched(fn) {
        this.formControlDirective.valueAccessor.registerOnTouched(fn);
    }
    setDisabledState(isDisabled) {
        this.formControlDirective.valueAccessor.setDisabledState(isDisabled);
    }
    writeValue(obj) {
        this.formControlDirective.valueAccessor.writeValue(obj);
    }
}
NgxGooglePlacesAutocompleteControlValueAccessor.propDecorators = {
    formControl: [{ type: Input }],
    formControlName: [{ type: Input }],
    formControlDirective: [{ type: ViewChild, args: [FormControlDirective, { static: true },] }]
};
