// @codious/ngx-google-places-autocomplete/esm2015/lib/ngx-google-places-autocomplete.component.js
import { __awaiter } from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Inject, Injector, Input, Output } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG } from './ngx-google-places-autocomplete.config';
import { NgxGooglePlacesAutocompleteControlValueAccessor } from './ngx-google-places-autocomplete.control-value-accessor';
export class NgxGooglePlacesAutocompleteComponent extends NgxGooglePlacesAutocompleteControlValueAccessor {
    constructor(changeDetectorRef, elementRef, injector, ngxGooglePlacesAutocompleteConfig) {
        super(injector);
        this.changeDetectorRef = changeDetectorRef;
        this.elementRef = elementRef;
        this.ngxGooglePlacesAutocompleteConfig = ngxGooglePlacesAutocompleteConfig;
        this.index = -1;
        this.predictions = [];
        this.show = false;
        this.placeholder = 'Enter a location';
        this.scriptLoaded = new EventEmitter();
        this.initialize().then(() => this.service = new google.maps.places.AutocompleteService());
    }
    onBlur() {
        this.reset(true);
    }
    onClick() {
        this.control.setValue(null);
    }
    onFocus() {
        this.reset(true, true);
    }
    onKeyDown(event) {
        if (!this.show)
            return;
        switch (event.key) {
            case 'ArrowDown':
                this.index += 1;
                if (this.index >= this.predictions.length)
                    this.index = this.predictions.length - 1;
                break;
            case 'ArrowUp':
                this.index -= 1;
                if (this.index < 0)
                    this.index = 0;
                break;
            case 'Escape':
                this.reset(true);
                break;
            case 'Enter':
                this.index !== -1 ? this.onMouseDown(this.predictions[this.index]) : this.reset(true);
                break;
        }
    }
    onModelChange(input) {
        if (!input) {
            this.reset();
            return;
        }
        const request = Object.assign({ input }, this.ngxGooglePlacesAutocompleteConfig.options);
        this.service.getPlacePredictions(request, (predictions, status) => {
            if (status !== google.maps.places.PlacesServiceStatus.OK) {
                this.reset();
                return;
            }
            this.predictions = predictions;
            this.show = true;
            this.changeDetectorRef.detectChanges();
        });
    }
    onMouseDown(prediction, event) {
        this.control.setValue(prediction.description, { emitViewToModelChange: false });
        if (event)
            this.dispatchEvent(event);
        this.reset(true);
    }
    dispatchEvent(event) {
        let customEvent;
        if (window.CustomEvent) {
            customEvent = new CustomEvent('submit', { bubbles: true, cancelable: true, detail: event });
        }
        else {
            customEvent = document.createEvent('CustomEvent');
            customEvent.initCustomEvent('submit', true, true, { data: event });
        }
        this.elementRef.nativeElement.firstElementChild.firstElementChild.form.dispatchEvent(customEvent);
    }
    initialize() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = window.google) === null || _a === void 0 ? void 0 : _a.maps)
                return new Promise((resolve) => resolve());
            return new Promise((resolve, reject) => {
                window['ngxGooglePlacesAutocompleteCallback'] = () => { this.scriptLoaded.emit(true); resolve(); };
                const script = document.createElement('script');
                script.async = true;
                script.defer = true;
                script.onerror = (err) => reject(err);
                script.src = `https://maps.googleapis.com/maps/api/js?callback=ngxGooglePlacesAutocompleteCallback&key=${this.ngxGooglePlacesAutocompleteConfig.key}&language=${this.ngxGooglePlacesAutocompleteConfig.language}&libraries=${this.ngxGooglePlacesAutocompleteConfig.libraries}&region=${this.ngxGooglePlacesAutocompleteConfig.region}`;
                script.type = 'text/javascript';
                document.body.appendChild(script);
            });
        });
    }
    reset(keep = false, show = false) {
        this.index = -1;
        if (!keep)
            this.predictions = [];
        this.show = show && this.predictions.length > 0;
    }
}
NgxGooglePlacesAutocompleteComponent.decorators = [
    { type: Component, args: [{
                providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: NgxGooglePlacesAutocompleteComponent, multi: true }],
                selector: 'ngx-google-places-autocomplete',
                template: "<div>\n  <input autocomplete=\"off\" (blur)=\"onBlur()\" (focus)=\"onFocus()\" [formControl]=\"control\" (keydown)=\"onKeyDown($event)\" (ngModelChange)=\"onModelChange($event)\" [placeholder]=\"placeholder\" type=\"text\" />\n  <button (click)=\"onClick()\" *ngIf=\"control.value\" type=\"button\">&times;</button>\n</div>\n<ul *ngIf=\"show\">\n  <li [class.selected]=\"i === index\" [innerHTML]=\"prediction.innerHTML\" (mousedown)=\"onMouseDown(prediction, $event)\" *ngFor=\"let prediction of predictions | highlight; let i = index\"></li>\n</ul>"
            },] }
];
NgxGooglePlacesAutocompleteComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG,] }] }
];
NgxGooglePlacesAutocompleteComponent.propDecorators = {
    placeholder: [{ type: Input }],
    scriptLoaded: [{ type: Output }]
};
