// @codious/ngx-google-places-autocomplete/esm2015/lib/ngx-google-places-autocomplete.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { NgxGooglePlacesAutocompleteComponent } from './ngx-google-places-autocomplete.component';
import { NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG } from './ngx-google-places-autocomplete.config';
import { NgxGooglePlacesAutocompletePipe } from './ngx-google-places-autocomplete.pipe';
export class NgxGooglePlacesAutocompleteModule {
    static forRoot(ngxGooglePlacesAutocompleteConfig) {
        return {
            ngModule: NgxGooglePlacesAutocompleteModule,
            providers: [{ provide: NGX_GOOGLE_PLACES_AUTOCOMPLETE_CONFIG, useValue: ngxGooglePlacesAutocompleteConfig }]
        };
    }
}
NgxGooglePlacesAutocompleteModule.decorators = [
    { type: NgModule, args: [{
                declarations: [NgxGooglePlacesAutocompleteComponent, NgxGooglePlacesAutocompletePipe],
                exports: [NgxGooglePlacesAutocompleteComponent],
                imports: [CommonModule, ReactiveFormsModule]
            },] }
];
